cloudinit/config/cc_ubuntu_advantage.py

```diff
@@ -288,35 +294,78 @@ def configure_ua(token, enable=None):
-        error = str(e).replace(token, REDACTED)
-        msg = f"Failure attaching Ubuntu Advantage:\n{error}"
+        err = str(e).replace(token, REDACTED)
+        msg = f"Failure attaching Ubuntu Advantage:\n{err}"
         util.logexc(LOG, msg)
         raise RuntimeError(msg) from e
 
-    enable_errors = []
-    for service in enable:
-        try:
-            cmd = ["ua", "enable", "--assume-yes", service]
-            subp.subp(cmd, capture=True)
-        except subp.ProcessExecutionError as e:
-            if re.search("is already enabled.", str(e)):
-                LOG.debug('Service "%s" already enabled.', service)
-            else:
-                enable_errors.append((service, e))
+    # Enable services
+    if not enable:
+        return
+    cmd = ["ua", "enable", "--assume-yes", "--format", "json", "--"] + enable
+    try:
+        enable_stdout, _ = subp.subp(cmd, capture=True, rcs={0, 1})
+    except subp.ProcessExecutionError as e:
+        raise RuntimeError(
+            f"Error while enabling service(s): {_format_services(enable)}"
+        ) from e
+
+    try:
+        enable_resp = json.loads(enable_stdout)
+    except json.JSONDecodeError as e:
+        raise RuntimeError(f"UA response was not json: {enable_stdout}") from e
+
+    # At this point we were able to load the json response from UA. This
+    # response contains a list of errors under the key 'errors'. E.g.
+    #
+    #   {
+    #      "errors": [
+    #        {
+    #           "message": "UA Apps: ESM is already enabled ...",
+    #           "message_code": "service-already-enabled",
+    #           "service": "esm-apps",
+    #           "type": "service"
+    #        },
+    #        {
+    #           "message": "Cannot enable unknown service 'asdf' ...",
+    #           "message_code": "invalid-service-or-failure",
+    #           "service": null,
+    #           "type": "system"
+    #        }
+    #      ]
+    #   }
+    #
+    # From our pov there are two type of errors, service and non-service
+    # related. We can distinguish them by checking if `service` is non-null
+    # or null respectively.
+
+    try:
+        from uaclient.messages import ALREADY_ENABLED  # noqa
+
+        UA_MC_ALREADY_ENABLED = ALREADY_ENABLED.name
+    except ImportError:
+        UA_MC_ALREADY_ENABLED = "service-already-enabled"
+
+    enable_errors: List[dict] = []
+    for err in enable_resp.get("errors", []):
+        if err["message_code"] == UA_MC_ALREADY_ENABLED:
+            LOG.debug("Service `%s` already enabled.", err["service"])
+            continue
+        enable_errors.append(err)
 
     if enable_errors:
-        for service, error in enable_errors:
-            msg = 'Failure enabling "{service}":\n{error}'.format(
-                service=service, error=str(error)
-            )
+        error_services: List[str] = []
+        for err in enable_errors:
+            if err["service"] is not None:
```

Comment:
Let's be definsive here and use `err.get("service")` in case the error does not contain a "service" key at some point in the future.

Reply:
done